### PR TITLE
Add embedding pooling kernel

### PR DESCRIPTION
**Pull request by z52527**
## Description
#214 
This PR adds a new embedding pooling kernel to address performance issues with `DynamicEmbeddingBagFunction`.

### Core Implementation
- `embedding_pooling_kernel.py`: Triton kernels (forward + backward)
- `embedding_pooling.py`: PyTorch autograd integration
- `test_embedding_pooling.py`: Unified test suite

## Usage Example

```python
from embedding_pooling import embedding_pooling

# Get embeddings from DynamicEmbeddingFunctionV2
embeddings = DynamicEmbeddingFunctionV2(...)  # [total_embs, dim]

# Apply pooling
pooled = embedding_pooling(embeddings, offsets, "mean")  # [batch, dim]
```
## Performance

| Test Case       | Config (Batch/Dim/Len) | Fwd-Triton | Fwd-PyTorch | Bwd-Triton | Bwd-PyTorch |
|:----------------|:--------------------|------------|-------------|------------|-------------|
| Small segments  | 100 / 128 / 10      | 2.752 μs | 21.76 μs | 3.2 μs | 7.808 μs |
| Medium segments | 1000 / 256 / 50     | 43.073 μs | 75.873 μs | 103.745 μs | 134.625 μs |
| Large segments  | 500 / 512 / 100     | 71.936 μs | 128.160 μs | 213.314 μs | 268.451 μs |
| Many segments   | 10000 / 128 / 20    | 70.144 μs  | 135.458 μs | 216.353 μs | 261.026 μs |
| Mixed lengths   | 1000 / 128 / None   | 28.352 μs | 52.195 μs | 47.681 μs | 63.073 μs |

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/rapids-repo-template/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    pooling_mode: str = "mean"
+) -> torch.Tensor:
+    """
+    Args:
+        embeddings: [total_embeddings, embedding_dim] - All embeddings
+        offsets: [num_segments + 1] - Segment boundaries
+        pooling_mode: "sum" or "mean"
+    Returns:
+        pooled: [num_segments, embedding_dim] - Pooled embeddings
+    """
```
Comment (shijieliu) [inline corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling.py:128]: could you add `assert pooling_mode in ['mean', 'sum']` here? We can add some check for the input tensor shape as well.

File: corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_autograd.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            grad_input_ptr=grad_embeddings,
+            embedding_dim=emb_dim,
+            num_segments=num_segs,
+            pooling_mode=mode,
+        )  
+
+        return grad_embeddings, None, None
+
+
+def embedding_pooling_with_grad(
```
Comment (shijieliu) [inline corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_autograd.py]: how about `embedding_pooling`

File: corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+            pooled = PoolingFunction.apply(embeddings, offsets, "mean")
+            loss = pooled.sum()
+            loss.backward()
+        torch.cuda.synchronize()
+        total_time = (time.time() - start) / num_iters * 1000
+        
+        # Measure forward only
+        start = time.time()
+        for _ in range(num_iters):
+            embeddings = torch.randn(total, dim, device='cuda')
```
Comment (shijieliu) [inline corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py]: I think it's better to move the initialization of embedding out of the benchmark loop. And it's better to use event to calculate the time.

File: corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling.py (hunk 1 above)
Comment (z52527) [reply corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling.py:128]: OK, I was asserting these in the forward function, but now I've moved them over here.

File: corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        triton.Config({"BLOCK_D": 256, "BLOCK_N": 128}, num_warps=8),
+        triton.Config({"BLOCK_D": 512, "BLOCK_N": 16}, num_warps=4),
+        triton.Config({"BLOCK_D": 512, "BLOCK_N": 32}, num_warps=8),
+        triton.Config({"BLOCK_D": 512, "BLOCK_N": 64}, num_warps=8),
+        triton.Config({"BLOCK_D": 512, "BLOCK_N": 128}, num_warps=8),
+    ],
+    key=["embedding_dim", "num_segments"],
+)
+@triton.jit
+def pooling_backward_kernel(
```
Comment (shijieliu) [inline corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py:125]: if we dont use this one, I think we can remove it?

File: corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+    pooling_backward_kernel_2d,
+    pooling_parallel_reduce_kernel,
+)
+
+# ============================================================================
+# Forward and backward: Only for testing
+# ============================================================================
+
+
+def embedding_pooling_forward_triton(
```
Comment (shijieliu) [inline corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py]: why do we need this? I think we need to test function in `embedding_pooling.py`

File: corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+    if pooling_mode == "sum":
+        output.scatter_add_(0, seg_ids.unsqueeze(1).expand(-1, dim), embeddings)
+    elif pooling_mode == "mean":
+        output.scatter_add_(0, seg_ids.unsqueeze(1).expand(-1, dim), embeddings)
+        output = output / lengths.unsqueeze(1).clamp(min=1)
+
+    return output
+
+
+def embedding_pooling_backward_torch(
```
Comment (shijieliu) [inline corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py]: I think the reference should be torch autograd backward not our implementation.

File: corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py (hunk 4 above)
Comment (z52527) [reply corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py:125]: In some cases, the 1D grid kernel performs better, so I've kept it as an option. However, the difference isn't particularly significant. Do you think it would be better to remove it?

File: corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        triton.Config({"BLOCK_D": 256, "BLOCK_N": 16}, num_warps=4),
+        triton.Config({"BLOCK_D": 256, "BLOCK_N": 32}, num_warps=4),
+        triton.Config({"BLOCK_D": 256, "BLOCK_N": 128}, num_warps=8),
+        triton.Config({"BLOCK_D": 256, "BLOCK_N": 32}, num_warps=8),
+        triton.Config({"BLOCK_D": 512, "BLOCK_N": 16}, num_warps=4),
+        triton.Config({"BLOCK_D": 512, "BLOCK_N": 32}, num_warps=8),
+        triton.Config({"BLOCK_D": 512, "BLOCK_N": 64}, num_warps=8),
+        triton.Config({"BLOCK_D": 512, "BLOCK_N": 128}, num_warps=8),
+    ],
+    key=["embedding_dim", "num_segments"],
```
Comment (geoffreyQiu) [inline corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py]: Using `num_segments` as tuning key directly might not be efficient. I think it'll cause tuning for every new #segs (need to check by profiler). Since our `BLOCK_N` is at least 16-mulpitle, do you think we can add an extra `autotune_num_segments` rounding by 16? Only if it does not harm performance. And it is only an inference issue, since in training the batch size is stable.

There is an example [link1](https://github.com/meta-recsys/generative-recommenders/blob/main/generative_recommenders/ops/triton/triton_hstu_attention.py#L1500), [link2](https://github.com/meta-recsys/generative-recommenders/blob/main/generative_recommenders/ops/triton/triton_hstu_attention.py#L1529-L1530), [link3](https://github.com/meta-recsys/generative-recommenders/blob/main/generative_recommenders/ops/triton/triton_hstu_attention.py#L2990-L2991). Just refer to [power_of_2](https://github.com/meta-recsys/generative-recommenders/blob/main/generative_recommenders/common.py#L282), and no need for runtime/static stuff.

Comment (z52527) [reply corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py]: You mean that autotune_num_segments is calculated by ‘power_of_2’ before entering the kernel at forward and backward, and then passed into the kernel, correct?

File: corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py (hunk 5 above)
Comment (z52527) [reply corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py]: Done.

File: corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py (hunk 6 above)
Comment (z52527) [reply corelib/dynamicemb/benchmark/embedding_pooling/test_embedding_pooling.py]: Done. Autograd gradcheck need FP64 but our kernel does not support that datatype. I compare our implementation to PyTorch Autograd result now.

File: corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py (hunk 7 above)
Comment (geoffreyQiu) [reply corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py]: Right, and triton.autotune will take it as the key. But I'm not sure rounding to whether pow2 or 16-multiple. 
16-multiple should have little impact on performance, but pow2 can reduce the chances of autotuning greatly. Need some benchmark to decide.

Comment (z52527) [reply corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py]: I think blocksize selection (powers of 2 or multiples of 16) doesn’t have much impact on performance. The bottleneck is the frequent  D2H during backward, which cannot be avoided. So powers of 2 are sufficient.

File: corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py (hunk 4 above)
Comment (shijieliu) [reply corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py:125]: For this PR I prefer code simplicty over performance.

Comment (z52527) [reply corelib/dynamicemb/benchmark/embedding_pooling/embedding_pooling_kernel.py:125]: Done.

**Review by geoffreyQiu — APPROVED** [review]
LGTM
